Fix the prof thread_name reference in prof_recent dump.

interwq opened the pull request:
As pointed out in #2434, the thread_name in prof_tdata_t was changed in #2407. This also requires an update for the prof_recent dump, specifically the emitter expects a "char **" which is fixed in this commit.

Also, add the prof_sys_thread_name feature in the prof_recent unit test, so that the combination of the two features is covered.  Verified that it catches the issue being fixed in this PR.

As a precaution, also went through all usage containing "thread_name" in the code base to make sure no similar issue exists.

Fixes #2434.

interwq committed:
Fix the prof thread_name reference in prof_recent dump.

As pointed out in #2434, the thread_name in prof_tdata_t was changed in #2407.
This also requires an update for the prof_recent dump, specifically the emitter
expects a "char **" which is fixed in this commit.

cemeyer reviewed (approved):
The proposed fix looks good to me.

interwq committed:
Add config detection for JEMALLOC_HAVE_PTHREAD_SET_NAME_NP.

and use it on the background thread name setting.

interwq committed:
Add the prof_sys_thread_name feature in the prof_recent unit test.

This tests the combination of the prof_recent and thread_name features.
Verified that it catches the issue being fixed in this PR.

Also explicitly set thread name in test/unit/prof_recent.  This fixes the name
testing when no default thread name is set (e.g. FreeBSD).
